lib/elm/compiler.rb

```diff
@@ -1,43 +1,53 @@
 require 'elm/compiler/exceptions'
 require 'open3'
 require 'tempfile'
-require 'mkmf'
 
 module Elm
   class Compiler
-    class << self
-      def compile(elm_files, output_path = nil)
-        fail ExecutableNotFound unless elm_executable_exists?
-
-        if output_path
-          elm_make(elm_files, output_path)
-        else
-          compile_to_string(elm_files)
-        end
-      end
+    def self.compile(*args)
+      new(*args).compile
+    end
+
+    def initialize(elm_files, output_path: nil, elm_make_path: nil)
+      @elm_files = elm_files
+      @output_path = output_path
+      @elm_make_path = elm_make_path
+    end
+
+    attr_reader :elm_files, :output_path, :elm_make_path
 
-      private
+    def compile
+      fail ExecutableNotFound unless elm_executable_exists?
 
-      def elm_executable_exists?
-        !find_executable0('elm-make').nil?
+      if output_path
+        to_file
+      else
+        to_s
       end
+    end
 
-      def compile_to_string(elm_files)
-        output = ''
+    private
 
-        Tempfile.open(['elm', '.js']) do |tempfile|
-          elm_make(elm_files, tempfile.path)
-          output = File.read tempfile.path
-        end
+    def elm_executable_exists?
+      File.exist?(elm_executable)
```

Comment:
It would be safer to use `File#executable?`, that way you can ensure that the file exists AND it's an executable.

```ruby
def elm_executable_exists?
  File.executable?(elm_executable)
end
```